pkg/agent/client/client.go

```diff
@@ -120,7 +120,7 @@ func (c *client) FetchUpdates(ctx context.Context, req *node.FetchX509SVIDReques
 	// Close the stream whether there was an error or not
 	stream.CloseSend()
 	if err != nil {
-		// TODO: should we try to create a new stream?
+		c.Release()
```

Comment:
What kind of errors might cause us to reach this point? Could any of them be recoverable i.e. might work if we asked the same server again? Similar question for below, but perhaps with a slightly different answer

Reply:
The errors that reach the point below (line 136) are those coming from the [server handler](https://github.com/spiffe/spire/blob/51e2e96f0fd8c2e612f884f9015cfca3510928af/pkg/server/endpoints/node/handler.go#L197).

Some of them does not seem to work if asked to the same server again, for example:

```
return status.Error(codes.ResourceExhausted, err.Error())
```


On the other hand, I'm not sure about errors reaching line 122 honestly. However, my intention here was to make sure that any failure leads to a new connection. So in case something unexpected happens on the server, the agent is able to recycle the connection and reach another server.

I found this particular issue difficult to reproduce so I did more focus on the second part.

What do you think? Happy to receive suggestions

Reply:
It is very difficult to reproduce these sorts of issues. This is the kind of thing that will surface and cause cascading failures in a large production environment when things go bad.

Essentially, it boils down to not beating your head against the wall, trying the same failing request over and over again to the same server and instead try reconnecting. Assuming a production environment with a scaled-up SPIRE server deployment behind a load balancer, there's a better chance that reconnecting will give you a good server, unless all servers are bad (in which case you have bigger issues anyway).

Reply:
Yea, wasn't suggesting that we _shouldnt_ reconnect. Ultimately, I was a little worried about small transient failures triggering stampeding herd of agent reconnects, because 1) tls negotiation isn't free at scale and 2) this issue https://github.com/spiffe/spire/issues/728

I got some rough numbers from @ZymoticB and @APTy, ran some back of the envelope calculation and think that we're probably OK and perhaps trying to recover before redialing is a premature optimization.

Reply:
I would invest in load shedding on the server side before I invested in recovering on the client side.

Exponential backoff+jitter on the client side and making rejecting new connections on the server side under load cheap will solve the stampeding herd concern IMO.